feat!: entrypoint cleanup

Thunkar opened the pull request:
This PR renames all `...MethodOptions` to `...InteractionOptions`. These are the collection of options that allow turning an `Interaction` into an `ExecutionPayload` the wallet can actually process. Also the convenience methods used to convert between the two are renamed from `sanitize...Option` to `to...Option`.

This PR introduces the concept of specialized entrypoints for different contracts. Our account contracts provide options to pay for the fee themselves via fee juice, or rely on an external method. This was previously achieved by a combination of lower level flags (`endSetup`, `isFeePayer`). This new system avoids inconsistent combinations.

Finally, the `FeeJuicePaymentMethod` has been removed, since it wasn't usable by apps and with the new flags the wallet can just command an account contract (that supports it) to pay for the fee with its own balance.

nventuro commented on yarn-project/entrypoints/src/account_entrypoint.ts:
```diff
@@ -6,9 +6,40 @@ import { HashedValues, TxContext, TxExecutionRequest } from '@aztec/stdlib/tx';
 import { ExecutionPayload } from './payload.js';
 
+/**
+ * An account contract can optionally pay for the fee of the transactions that go through it, but in order
+ * to do that it has to perform certain operations depending on the payment method.
+ *  - Directly using its fee juice balance: it has to set itself as fee payer and end the setup phase.
+ *  - In fee juice, but using a combination of its balance and a claim *in the same transaction*:
+ *    only set itself as fee payer
+ *  - The fee is handled by other calls in the tx: do nothing
+ * This options convey in which scenario we are, so the contract can act accordingly
+ */
+export enum AccountFeePaymentMethodOptions {
+  FEE_JUICE = 1,
+  FEE_JUICE_WITH_CLAIM = 2,
+  EXTERNAL = 3,
+}
```
```suggestion
 * The mechanism via which an account contract will pay for a transaction in which it gets invoked.
 */
export enum AccountFeePaymentMethodOptions {
  /**
   * Used to make the account contract publicly pay for the transaction with its own fee juice balance,
   * **which it must already have prior to this transaction**. 
   * 
   * The contract will set itself as the fee payer and end the setup phase.
   */
  FEE_JUICE = 1,
  /**
   * Used to make the account contract publicly pay for the transaction with its own fee juice balance
   * **which is being claimed in the same tranasction**. 
   * 
   * The contract will set itself as the fee payer but not end setup phase - this is done by the Fee Juice
   * contract after enqueuing a public call, which unlike most public calls is whitelisted by the nodes
   * to be executable during the setup phase.
   */
  FEE_JUICE_WITH_CLAIM = 2,
  /**
   * Signals that some other contract is in charge of paying the fee, nothing needs to be done.
   */
  EXTERNAL = 3,
}
```

I like using docstrings this way because it's easy to do e.g. hover and quicky get a sense of what's going on. I did some light rephrasing including some bits that I thought would be relevant for a reader that doesn't know a lot about the details of what's going on.

What do you think about renaming `FEE_JUICE` to `PREEXISTING_FEE_JUICE`? That should make people ask themselves very important questions.

nventuro commented on noir-projects/aztec-nr/aztec/src/authwit/account.nr:
```diff
@@ -18,6 +18,15 @@ impl<Context> AccountActions<Context> {
     }
 }
 
+pub struct FeePaymentMethodEnum {
```
```suggestion
/// Noir representation of the `AccountFeePaymentMethodOptions` in the TypeScript `entrypoint` package.
pub struct FeePaymentMethodEnum {
```

Note that in js we call this struct `AccountFeePaymentMethodOptions`, it'd be nice for them to share the name.

Thunkar replied:
Renamed both the same for consistency, changed `FEE_JUICE` option to `CURRENT_FEE_JUICE_BALANCE`. WDYT?

Thunkar replied:
Also shuffled the options around so EXTERNAL gets assigned value 0...just in case TS messes up. There shouldn't be a difference. but lines up better with "do nothing unless explicitly asked"

nventuro replied:
I think 'preexisting' makes it much more obvious that it is wrong if you're thinking about claiming and using it immediately. 'current' is a bit vague - what'd be not current? Surely not future, so past? It turns out it actually _is_ the past balance, not the 'current' balance. Idk it gets into annoying timing semantics. I prefer 'preexisting' personally.

Thunkar replied:
Well, it's current because you haven't claimed YET. It will happen in the same tx, but strictly speaking you're paying with a combination of your current (right now) balance and the sum of the claim that will be resolved later in public

Thunkar replied:
🤷‍♂️

nventuro commented on boxes/boxes/vanilla/app/embedded-wallet.ts:
```diff
@@ -81,19 +82,24 @@ export class EmbeddedWallet extends BaseWallet {
-      paymentMethod = new SponsoredFeePaymentMethod(
+      const sponsoredFPCContract =
+        await EmbeddedWallet.#getSponsoredPFCContract();
+      walletFeePaymentMethod = new SponsoredFeePaymentMethod(
         sponsoredFPCContract.instance.address
       );
+      accountFeePaymentMethodOptions = AccountFeePaymentMethodOptions.EXTERNAL;
```
Would `walletFeePaymentMethod` ever be set of the option is not external?

Thunkar replied:
It should always be set to *something*...but just in case, I shuffled the options around and assigned the "EXTERNAL" option to value 0 in the enum. It lines up better with "do nothing unless explicitly stated"

Thunkar replied:
Yes, if it's set to "fee juice with claim"

nventuro reviewed (commented):
Lovely work, thanks a lot for improving these things.

benesjan commented on noir-projects/aztec-nr/aztec/src/authwit/account.nr:
```diff
@@ -18,6 +18,18 @@ impl<Context> AccountActions<Context> {
     }
 }
 
+pub struct AccountFeePaymentMethodOptionsEnum {
```
```suggestion
// See AccountFeePaymentMethodOptions enum in Aztec.js for docs:
// https://github.com/AztecProtocol/aztec-packages/blob/next/yarn-project/entrypoints/src/account_entrypoint.ts
pub struct AccountFeePaymentMethodOptionsEnum {
```

If we don't want to copy the docs would point devs to where they exists. The link potentially becoming stale is not great though so maybe just copying it would be better. But I think I would do it once everything is more stable.

benesjan commented on yarn-project/entrypoints/src/account_entrypoint.ts:
```diff
@@ -6,9 +6,51 @@ import { HashedValues, TxContext, TxExecutionRequest } from '@aztec/stdlib/tx';
+  PREEXISTING_FEE_JUICE = 1,
+  /**
+   * Used to make the account contract publicly pay for the transaction with its own fee juice balance
+   * **which is being claimed in the same tranasction**.
```
```suggestion
   * **which is being claimed in the same transaction**.
```

BTW docs now run spellcheck on build so if this was included in docs you would get a CI failure (like me [here](https://aztecprotocol.slack.com/archives/C08HLF3J941/p1759216431721869)).

So recommend installing [the spellcheck extension](https://marketplace.cursorapi.com/items/?itemName=streetsidesoftware.code-spell-checker).

Thunkar replied:
Installed!

benesjan commented on boxes/boxes/vanilla/app/embedded-wallet.ts:
```diff
@@ -81,19 +82,24 @@ export class EmbeddedWallet extends BaseWallet {
     } else {
-      // The transaction includes fee payment method, so we check if we are the fee payer for it (it could only be FeeJuiceWithClaim)
-      isFeePayer = from.equals(userFeeOptions.embeddedPaymentMethodFeePayer);
+      // The transaction includes fee payment method, so we check if we are the fee payer for it
+      // (this can only happen if the embedded payment method is FeeJuiceWithClaim)
+      accountFeePaymentMethodOptions = from.equals(
+        userFeeOptions.embeddedPaymentMethodFeePayer
+      )
+        ? AccountFeePaymentMethodOptions.FEE_JUICE_WITH_CLAIM
+        : AccountFeePaymentMethodOptions.EXTERNAL;
     }
```
I feel like the behavior here is reversed because we have the nice `AccountFeePaymentMethodOptions` enums but instead of passing those on the input and based on that setting the options we pass on the input the options and based derive the enum.

Would it be so complicated to swap the direction here?

benesjan replied:
Discussed in DMs:

<img width="572" height="455" alt="Image" src="https://github.com/user-attachments/assets/b39e5ec7-6e71-4312-b634-3ecfdfd37b5f" />

benesjan commented on yarn-project/aztec.js/src/contract/interaction_options.ts:
```diff
@@ -23,7 +23,7 @@ export type FeeEstimationOptions = {
  */
 export type FeePaymentMethodOption = {
   /** Fee payment method to embed in the interaction */
```
```suggestion
  /**
   * Fee payment method to embed in the interaction. If not set, defaults to the account paying with its own fee juice
   * balance.
   */
```

Thunkar replied:
This is unfortunately untrue, since the wallet can decide which method is the default!

benesjan commented on yarn-project/aztec.js/src/contract/interaction_options.ts:
```diff
@@ -35,7 +35,6 @@ export type GasSettingsOption = {
-// docs:start:user_fee_options
 /** Fee options as set by a user. */
 export type InteractionFeeOptions = GasSettingsOption & FeePaymentMethodOption;
 // docs:end:user_fee_options
```
```suggestion
```

benesjan commented on yarn-project/aztec.js/src/fee/fee_juice_payment_method_with_claim.ts:
```diff
@@ -3,34 +3,25 @@ import { Fr } from '@aztec/foundation/fields';
-  _branding: 'AppConfigurableFeePaymentMethod';
-}
+import type { FeePaymentMethod } from './fee_payment_method.js';
 
 /**
  * Pay fee directly with Fee Juice claimed on the same tx.
  */
```
```suggestion
/**
 * Pay fee directly with Fee Juice claimed in the same tx. Claiming consumes an L1 to L2 message that "contains"
 * the fee juice bridged from L1.
 */
```
Unrelated to your PR but found the docs here a bit too concise.

benesjan commented on yarn-project/aztec.js/src/fee/fee_juice_payment_method_with_claim.ts:
```diff
@@ -3,34 +3,25 @@ import { Fr } from '@aztec/foundation/fields';
 
   /**
    * Creates an execution payload to pay the fee in Fee Juice.
    * @returns An execution payload that just contains the claim function call.
```
```suggestion
   * @returns An execution payload that just contains the `claim_and_end_setup` function call.
```

benesjan commented on yarn-project/aztec.js/src/wallet/account_entrypoint_meta_payment_method.ts:
```diff
@@ -15,43 +17,57 @@ import type { FeePaymentMethod } from '../fee/fee_payment_method.js';
   async getExecutionPayload(): Promise<ExecutionPayload> {
     // Get the execution payload for the fee, it includes the calls and potentially authWitnesses
-    const { calls: feeCalls, authWitnesses: feeAuthwitnesses } = await this.paymentMethod.getExecutionPayload();
+    // It can be emtpy because the account might attempt to pay for the tx fee using it's own
```
```suggestion
    // It can be empty because the account might attempt to pay for the tx fee using its own
```

benesjan commented on yarn-project/aztec.js/src/wallet/account_entrypoint_meta_payment_method.ts:
```diff
@@ -15,43 +17,57 @@ import type { FeePaymentMethod } from '../fee/fee_payment_method.js';
  * It works by rerouting the provided fee payment method through the account's entrypoint,
- * which sets itself as fee payer.
+ * which sets itself as fee payer. If no payment method is provided, it is assumed the
+ * account will pay with its own fee juice balance.
```
Unrelated but it's pretty cool that this is even possible. Would be a pity to have the drop fee abstraction (at least that's what I understood is being considered).

Thunkar replied:
Agreed, but after dealing with all the setup/teardown revertible/non-revertible stuff I wonder if the complexity is worth it...

benesjan commented on yarn-project/aztec.js/src/wallet/base_wallet.ts:
```diff
@@ -54,17 +53,10 @@ import type {
-   * contract executing the transaction should end the setup phase
-   */
-  endSetup: boolean;
+  /** Wallet provided fee payment method, in case the tx doesn't include one */
```
```suggestion
  /**
   * A wallet-provided fallback fee payment method that is used only if the transaction that is being constructed
   * doesn't already include one
   */
```
I found the original a bit confusing because the term "transaction" in blockchains in generally refers to a signed immutable object. Hence the wallet having the ability to modify it is strange. Pardon my autism

benesjan commented on yarn-project/entrypoints/src/interfaces.ts:
```diff
@@ -5,29 +5,12 @@ import type { TxExecutionRequest } from '@aztec/stdlib/tx';
-   * FeeJuiceWithClaim the account contract is the fee payer, but the end of the setup is handled by the FeeJuice contract.
-   */
-  endSetup: boolean;
+/** Information on the connected chain */
```
```suggestion
/**
 * Information on the connected chain. Used by wallets when constructing transactions to protect against replay
 * attacks.
 */
```

benesjan commented on yarn-project/entrypoints/src/interfaces.ts:
```diff
@@ -39,13 +22,13 @@ export interface EntrypointInterface {
     exec: ExecutionPayload,
     gasSettings: GasSettings,
-    options: TxExecutionOptions,
+    options?: any,
```
Is there a plan to re-introduce some kind of type here?

Thunkar replied:
This is an unfortunate limitation of typescript itself. This interface essentially says: "an entrypoint has two mandatory params and the other two depend on contract implementation"

If people come up with alternative account contracts, we don't want to force our options on them.

benesjan reviewed (approved):
Very nice!

The suggestions I made that are not a direct docs improvements are mostly just talking points for the future. This PR is indeed a strict improvement to what's on `next` so doesn't make sense to not merge this ASAP.

Thanks

Thunkar committed:
feat!: entrypoint cleanup

This PR renames all `...MethodOptions` to `...InteractionOptions`. These are the collection of options that allow turning an `Interaction` into an `ExecutionPayload` the wallet can actually process. Also the convenience methods used to convert between the two are renamed from `sanitize...Option` to `to...Option`.

This PR introduces the concept of specialized entrypoints for different contracts. Our account contracts provide options to pay for the fee themselves via fee juice, or rely on an external method. This was previously achieved by a combination of lower level flags (`endSetup`, `isFeePayer`). This new system avoids inconsistent combinations.

Finally, the `FeeJuicePaymentMethod` has been removed, since it wasn't usable by apps and with the new flags the wallet can just command an account contract (that supports it) to pay for the fee with its own balance.

Co-authored-by: Gregorio Juliana <gregojquiros@gmail.com>
